src/bananas.js

```diff
@@ -0,0 +1,185 @@
+                    node.toHashtag = function() {
+                        return transformFn(oldToHashtag());
+                    };
+                })();
+            }
+            children = node.getChildren();
+            for (i = 0; i < children.length; i++) {
+                queue.push(children[i]);
+                if (children[i].predicates && children[i].predicates.length) {
+                    predicates = children[i].predicates;
+                    for (j = 0; j < predicates.length; j++) {
+                        queue.push(predicates[j]);
+                    }
+                }
+            }
+        }
+        return parsedHashtags.toHashtag();
+    }
+
+    /*
+     * transforms bananas based on transformFn
+     */
+    function transform(input, transformFn) {
+        if (!input) { return input; }
+        input = getSymbols(input);
+        transformFn = transformFn || function (input) { return input; };
+        var state = OUTSIDE_BANANA,
+            strLen = input.length,
+            text = "",
+            currentReference = "";
+
+        for (var i = 0; i < strLen; i++) {
+            var current = input[i],
+                next = input[i+1];
+
+            if (state === OUTSIDE_BANANA) {
+                if (current === DELIMITER && next === DELIMITER) {
+                    text += DELIMITER;
+                    i++;
+                } else if (current === DELIMITER) {
+                    state = INSIDE_BANANA;
+                } else {
+                    text += current;
+                }
+            } else if (state === INSIDE_BANANA) {
+                if (current === DELIMITER && next === DELIMITER) {
+                    currentReference += DELIMITER;
+                    i++;
+                } else if (current === DELIMITER) {
+                    state = OUTSIDE_BANANA;
+                    text += transformFn(currentReference);
+                    currentReference = "";
+                } else if (next !== undefined){
+                    currentReference += current;
+                } else {
+                    // end of string, shouldn't happen, but will not
+                    // overestimate users or Vellum devs
+                    text += DELIMITER + currentReference;
+                    currentReference = "";
+                }
+            }
+        }
+
+        if (state === INSIDE_BANANA) {
+            // end of string, without the end delimiter
+            text += DELIMITER + currentReference;
+        }
+
+        return text;
+    }
+
+    function getSymbols(string) {
+        var index = 0;
+        var length = string.length;
+        var output = [];
+        for (; index < length - 1; ++index) {
+            var charCode = string.charCodeAt(index);
+            if (charCode >= 0xD800 && charCode <= 0xDBFF) {
+                charCode = string.charCodeAt(index + 1);
+                if (charCode >= 0xDC00 && charCode <= 0xDFFF) {
```

Comment:
what are the magic codepoint numbers? would it make sense to use named constants?